Homework 6

prometey-dev opened the pull request:
1. Added background loading of the list of movies using corrutines
2. Loading pictures via coil. First I added glide, then replaced it with coil. In coil, I loaded an image into a movie adapter using corrutines.
3. Added saving / restoring the state of the layout-manager to the screen with the list of films when turning the screen and when navigate to the screen with details.

![image](https://user-images.githubusercontent.com/6102319/101776773-860e7580-3b02-11eb-936d-d333dfddb96f.png)

![image](https://user-images.githubusercontent.com/6102319/101776906-b6561400-3b02-11eb-9353-0dcacc1ca6cf.png)

prometey-dev committed:
Add kotlin serialisation and coroutines to project

Commit:
Merge branch 'master' into homework_6

# Conflicts:
#	app/build.gradle

prometey-dev committed:
Add prepared asset and data serialization

prometey-dev committed:
Add glide to project

prometey-dev committed:
Load movie with corrutines and transfer movie parcelable to details fragment

prometey-dev committed:
Replace Glide to Coil

prometey-dev committed:
Add CardView instead parent ImageView to view_holder_movie.xml

prometey-dev committed:
Adding for CardView

prometey-dev committed:
Adding dimens for CardView

prometey-dev committed:
Remove unused classes

prometey-dev committed:
save/restore state layout manager for recycler

prometey-dev committed:
Calculate stars count, vote average divide 2

prometey-dev committed:
Loading pictures in the list of films in coil via ImageLoader on corrutines

prometey-dev committed:
Remove commented code

emartynov commented:
> 1. Added background loading of the list of movies using corrutines
> 2. Loading pictures via coil. First I added glide, then replaced it with coil. In coil, I loaded an image into a movie adapter using corrutines.
> 3. Added saving / restoring the state of the layout-manager to the screen with the list of films when turning the screen and when navigate to the screen with details.
> 
> ![image](https://user-images.githubusercontent.com/6102319/101776773-860e7580-3b02-11eb-936d-d333dfddb96f.png)
> 
> ![image](https://user-images.githubusercontent.com/6102319/101776906-b6561400-3b02-11eb-9353-0dcacc1ca6cf.png)

Do images of actor look wrong for you also?

emartynov commented on app/src/main/AndroidManifest.xml:
```diff
@@ -2,6 +2,9 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     package="ru.prometeydev.movie">
 
+    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
```
Why do you need this permission?

prometey-dev replied:
Yes, we do not need this permission at the moment. INTERNET permission is sufficient. I will delete it.

emartynov commented on app/src/main/java/ru/prometeydev/movie/FragmentMoviesDetails.kt:
```diff
@@ -45,19 +45,19 @@ class FragmentMoviesDetails : Fragment() {
 
         view.findViewById<TextView>(R.id.age_limit)
-            .text = context?.getString(R.string.age_limit, movie?.ageLimit)
+            .text = context?.getString(R.string.age_limit, movie?.minimumAge)
```
Fragment has own getString string

prometey-dev replied:
Ok, I will improve

emartynov commented on app/src/main/java/ru/prometeydev/movie/FragmentMoviesDetails.kt:
```diff
@@ -45,19 +45,19 @@ class FragmentMoviesDetails : Fragment() {
 
         view.findViewById<ImageView>(R.id.movie_logo)
-            .setImageResource(movie?.additional!!.pictureDrawable)
+            .load(movie?.backdrop)
```
Why not to check movie for null once and remove all `?` later?

prometey-dev replied:
Ok, I will improve

emartynov commented on app/src/main/java/ru/prometeydev/movie/FragmentMoviesList.kt:
```diff
@@ -44,27 +47,46 @@ class FragmentMoviesList : Fragment() {
-            return
+        recycler?.let {
+            arguments = Bundle().apply {
+                putParcelable(BUNDLE_RECYCLER_LAYOUT, it.layoutManager?.onSaveInstanceState())
```
Why on save instance state is not enough?

prometey-dev replied:
Because onSaveInstanceState occurs when the screen is rotated, and when you go to the next screen by clicking on the card, it is not called.

emartynov commented on app/src/main/java/ru/prometeydev/movie/FragmentMoviesList.kt:
```diff
@@ -1,27 +1,30 @@
 
     private var spanCount = VERTICAL_SPAN_COUNT
 
+    private var movies: List<Movie> = listOf()
```
Why store movies?

prometey-dev replied:
Yes, you don't need to store it. Download and bind immediately.

emartynov commented on app/src/main/java/ru/prometeydev/movie/data/adapters/MoviesAdapter.kt:
```diff
@@ -60,29 +66,39 @@ class MoviesAdapter(
+            reviewsCount.text = context.getString(R.string.reviews, movie.numberOfRatings)
+            ageLimit.text  = context.getString(R.string.age_limit, movie.minimumAge)
+
+            scope.launch {
```
This is dangerous - do you need to cancel scope when you reuse holder?

prometey-dev replied:
Yeah, I’ll fix it.

prometey-dev replied:
I decided to remove the use of corrutines here, since it doesn't seem to affect the speed. I will also remove ImageLoader from Coil.

emartynov commented on app/src/main/java/ru/prometeydev/movie/data/adapters/MoviesAdapter.kt:
```diff
@@ -60,29 +66,39 @@ class MoviesAdapter(
             }
         }
 
+        private suspend fun loadImage(url: String): Drawable? = withContext(Dispatchers.IO) {
```
Is this exercise since I saw something like this already in coil?

prometey-dev replied:
I think a simple call to load(url) will be enough.

emartynov reviewed (commented):
Amazing to see a almost live data!

I didn't understand the git submodule and log files in the repository.

A couple of major comments:
1. Extra permission
2. Handling image loading cancellation for the holder

And a couple of small comments and questions.

prometey-dev commented:
> I didn't understand the git submodule and log files in the repository.

It turned out by accident, I need to fix .gitignore.

prometey-dev committed:
Remove permission ACCESS_NETWORK_STATE from AndroidManifest.xml

prometey-dev committed:
Fix scale images actors

prometey-dev committed:
Remove property movies from FragmentMovieList

prometey-dev committed:
Add movie check null

prometey-dev committed:
Replace context.getString on getString at FragmentMoviesDetails

prometey-dev committed:
Remove use corrutines from MoviesAdapter

prometey-dev committed:
Improved saving / restoring the state of the layout manager

prometey-dev commented:
Fixed the size of the ImageView with the actors, and also added some fixes / improvements.

![image](https://user-images.githubusercontent.com/6102319/101880863-a2fb8500-3ba4-11eb-82a3-2bcb8de16db8.png)

emartynov commented on app/src/main/java/ru/prometeydev/movie/FragmentMoviesDetails.kt:
```diff
@@ -44,32 +44,38 @@ class FragmentMoviesDetails : Fragment() {
-        view.findViewById<TextView>(R.id.age_limit)
-            .text = context?.getString(R.string.age_limit, movie?.ageLimit)
+            view.findViewById<TextView>(R.id.age_limit)
+                    .text = getString(R.string.age_limit, it.minimumAge)
```
Minor - `it` is fine for one liner and it is harder to follow when it is used in many places. Consider to have something meaningful

prometey-dev replied:
Improve it

emartynov reviewed (approved):
LGTM

Commit:
Replace it to mov in let

emartynov commented on app/src/main/java/ru/prometeydev/movie/FragmentMoviesDetails.kt:
```diff
@@ -44,21 +44,21 @@ class FragmentMoviesDetails : Fragment() {
         loadData()
 
-        movie?.let {
+        movie?.let { mov ->
```
Why not to `movie`?

prometey-dev replied:
Because the private var movie is already there, so as not to get confused.

emartynov replied:
Sorry, for long reply, I would probably use something like `movieValue` or so.
